contrib/direnv

```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env bash
+
+# Look for the .config folder in the current directory and up
+find_prj_config() (
+	local old_pwd
+	while [[ $old_pwd != $PWD ]]; do
+		if [[ -d .config ]]; then
+			echo "$PWD"
+			return 0
+		fi
+		old_pwd=$PWD
+		cd ..
+	done
+	# We're at the top and didn't find anything
+	log_error "ERROR: could not find project root" >&2
+	return 1
+)
+
+{
+	: "${XDG_CONFIG_HOME:=${HOME}/.config}"
+	: "${XDG_CACHE_HOME:=${HOME}/.cache}"
+	: "${XDG_DATA_HOME:=${HOME}/.local/share}"
+	: "${XDG_STATE_HOME:=${HOME}/.local/state}"
+	: "${XDG_RUNTIME_DIR:=${HOME}/.local/state}"
+
+	: "${GIT_DIR:=$(git rev-parse --absolute-git-dir 2>/dev/null)}"
+
+	# PRJ_ROOT
+	_prj_root="${GIT_DIR%/.git}"
+	: "${PRJ_ROOT:=${_prj_root:=$(find_prj_config)}}"
+
+	# PRJ_CONFIG_HOME - always local to the project
+	: "${PRJ_CONFIG_HOME:=${PRJ_ROOT}/.config}"
+	mkdir -p "${PRJ_CONFIG_HOME}"
+
+	if [[ -z "${PRJ_ID:-}" && -f "${PRJ_CONFIG_HOME}/prj_id" ]]; then
+		PRJ_ID=$(<"${PRJ_CONFIG_HOME}/prj_id")
+	fi
+
+	# PRJ_RUNTIME_DIR - always local to the project
+	: "${PRJ_RUNTIME_DIR:=${PRJ_ROOT}/.run}"
+	mkdir -p "${PRJ_RUNTIME_DIR}"
+
+	# PRJ_CACHE_HOME - shared if PRJ_ID is set
+	if [[ -z "${PRJ_CACHE_HOME:-}" ]]; then
+		if [[ -n "${PRJ_ID:-}" ]]; then
+			PRJ_CACHE_HOME="${XDG_CACHE_HOME}/prj/${PRJ_ID}"
+		else
+			PRJ_CACHE_HOME="${PRJ_ROOT}/.cache"
+		fi
+	fi
+	mkdir -p "${PRJ_CACHE_HOME}"
+
+	# PRJ_DATA_HOME - shared if PRJ_ID is set
+	if [[ -z "${PRJ_DATA_HOME:-}" ]]; then
+		if [[ -n "${PRJ_ID:-}" ]]; then
+			PRJ_DATA_HOME="${XDG_DATA_HOME}/prj/${PRJ_ID}"
+		else
+			PRJ_DATA_HOME="${PRJ_ROOT}/.local/share"
+		fi
+	fi
+	mkdir -p "${PRJ_DATA_HOME}"
+
+	# PRJ_STATE_HOME - shared if PRJ_ID is set
+	if [[ -z "${PRJ_STATE_HOME:-}" ]]; then
+		if [[ -n "${PRJ_ID:-}" ]]; then
+			PRJ_STATE_HOME="${XDG_STATE_HOME}/prj/${PRJ_ID}"
+		else
+			PRJ_STATE_HOME="${PRJ_ROOT}/.local/state"
+		fi
+	fi
+	mkdir -p "${PRJ_STATE_HOME}"
+
+	# PRJ_PATH - shared if PRJ_ID is set
+	if [[ -z "${PRJ_PATH:-}" ]]; then
+		if [[ -n "${PRJ_ID:-}" ]]; then
+			PRJ_PATH="${HOME}/.local/bin/prj/${PRJ_ID}"
+		else
+			PRJ_PATH="${PRJ_ROOT}/.local/bin"
+		fi
+	fi
+	mkdir -p "${PRJ_PATH}"
+
+	# Git ignore all `PRJ_*` in the main checkout
+	# i.e. if we're not in a worktree, that is: GIT_DIR is not a file
+	if [[ -n "${GIT_DIR}" && -d "${GIT_DIR}" ]]; then
+		for varname in \
+			PRJ_CONFIG_HOME \
+			PRJ_CACHE_HOME \
+			PRJ_DATA_HOME \
+			PRJ_STATE_HOME \
+			PRJ_RUNTIME_DIR; do
+			[[ ! ${!varname} == ${PRJ_ROOT}* ]] && continue
+			grep "${!varname#${PRJ_ROOT}}" "$GIT_DIR/info/exclude" &>/dev/null && continue
+			cat >>"$GIT_DIR/info/exclude" <<<"${!varname#${PRJ_ROOT}}"
+			log_status Add to "${GIT_DIR#${PRJ_ROOT}/}/info/exclude": "${!varname#${PRJ_ROOT}}"
+		done
+	fi
+}
+
+unset -f find_prj_config
+
+PATH_add "${PRJ_PATH}"
+export PRJ_ROOT
+export PRJ_ID
+export PRJ_PATH
+export PRJ_CONFIG_HOME
+export PRJ_CACHE_HOME
+export PRJ_DATA_HOME
+export PRJ_STATE_HOME
+export PRJ_RUNTIME_DIR
```

Comment:
What is the difference between PRJ_DATA_HOME, PRJ_STATE_HOME and PRJ_RUNTIME_DIR?

Reply:
According to `XDG`, `DATA` is for potentially sharable data that is useful independent of the context of the concrete user environment. One example that comes to my mind would be blockchain state. A way to think about it would be a hypothetical user that works on mac and linux interchangeably but shares projects on a NFS. `DATA` should _not_ hold arch-specific data.

`STATE`, however, could be a stateful application file (these are rare these days), that don't immediately make sense to share, either because they are arch-specific or their lifetime is bound to the host machine. A better use in most cases would be `/tmp`, but still, maybe an application stores last startup configuration to resume on. The Paisano TUI/CLI is recording the last executed action there in order to facilitate manual debugging but also because the invocation sequence requires a file proxy at some point.

Reply:
`PRJ_RUNTIME_DIR` is usually just pids and sockets, maybe a runtime wal or such.

Reply:
Removed `PRJ_STATE_HOME` in 7dfd31d4d82c6ebdec54eface81dfb5cce689eab.

But I'm keeping `PRJ_RUNTIME_DIR` for deveshell services use case ("find their pids easily").

Reply:
@zimbatm would that be ok (keep PRJ_RUNTIME_DIR)? Would you like me to expand more on the use case?